vulture_os/toolkit/api_parser/netskope/netskope.py

```diff
@@ -114,45 +126,62 @@ def get_logs(self, since, to=timezone.now(), cursor=0):
         assert result['ok'] == 1, result['message']
         return result
 
-    def parse_log(self, log):
+    def parse_alert_log(self, log):
         if log['timestamp'] > self.upper_timestamp:
             self.upper_timestamp = log['timestamp']
         log['alert_type'] = log['alert_type'].lower()
         return json.dumps(log)
 
     def execute(self):
-        self.upper_timestamp = int(self.frontend.last_api_call.timestamp()) if self.frontend.last_api_call else 0
-        since = self.last_api_call or (datetime.utcnow() - timedelta(hours=24))
-        to = min(timezone.now(), since + timedelta(hours=24))
-        offset = 0
-
-        while offset%self.BULK_SIZE == 0:
-
-            response = self.get_logs(since, to, offset)
-
-            # Downloading may take some while, so refresh token in Redis
-            self.update_lock()
-
-            logs = response['result']
-            if len(logs) == 0:
-                logger.info(f"[{__parser__}][execute]: No more log to fetch. End of the parsing",
-                            extra={'frontend': str(self.frontend)})
-                break
-            offset += len(logs)
-
-            self.write_to_file([self.parse_log(l) for l in logs])
-
+        logtypes = ["alert"]
+        if self.netskope_get_page_logs:
+            logtypes.append("page")
+        if self.netskope_get_network_logs:
+            logtypes.append("network")
+        if self.netskope_get_application_logs:
+            logtypes.append("application")
+        for logtype in logtypes:
+            since = self.last_collected_timestamps.get(logtype, self.last_api_call) or (timezone.now() - timedelta(hours=24))
+            self.upper_timestamp = since.timestamp()
+            to = min(timezone.now(), since + timedelta(hours=24))
+            offset = 0
+            logs = []
+            log_ids = set()
+
+            while offset%self.BULK_SIZE == 0:
+
+                response = self.get_logs(since, to, offset, logtype)
+
+                # Downloading may take some while, so refresh token in Redis
+                self.update_lock()
+
+                new_logs = response['result']
+                if len(new_logs) == 0:
+                    logger.info(f"[{__parser__}][execute]: No more log to fetch. End of the parsing",
+                                extra={'frontend': str(self.frontend)})
+                    break
+                # Remove duplicated logs
+                log_ids = log_ids.union(set([log.get("_event_id", "") for log in logs]))
```

Comment:
This is **not** better: you're still recreating the list from all the logs present in the 'logs' whole list.
Moreover: if some logs don't have an _event_id, all following logs without one will get dropped, even when different!